capture/capture_aes.py

```diff
@@ -0,0 +1,347 @@
+    remaining_num_traces = capture_cfg.num_traces
+    with tqdm(total=remaining_num_traces, desc="Capturing", ncols=80, unit=" traces") as pbar:
+        while remaining_num_traces > 0:
+            # Arm the scope.
+            scope.arm()
+
+            # Trigger encryption.
+            if capture_cfg.batch_mode:
+                # Batch mode.
+                if capture_cfg.capture_mode == "aes_random":
+                    # Fixed key, random plaintexts.
+                    ot_aes.encrypt_batch(
+                        capture_cfg.num_segments.to_bytes(4, "little"))
+                else:
+                    # Fixed vs random key test.
+                    ot_aes.encrypt_fvsr_key_batch(
+                        capture_cfg.num_segments.to_bytes(4, "little"))
+            else:
+                # Non batch mode.
+                ot_aes.encrypt(text)
+
+            # Capture traces.
+            waves = scope.capture_and_transfer_waves()
+            assert waves.shape[0] == capture_cfg.num_segments
+            # Generate reference crypto material and store trace.
+            for i in range(capture_cfg.num_segments):
+                text, key, ciphertext, sample_fixed = generate_ref_crypto(
+                    sample_fixed = sample_fixed,
+                    mode = capture_cfg.capture_mode,
+                    key = key,
+                    plaintext = text,
+                    key_length = capture_cfg.key_len_bytes,
+                    text_length = capture_cfg.text_len_bytes
+                )
+                # Sanity check retrieved data (wave).
+                assert len(waves[i, :]) >= 1
+                # Store trace into database.
+                project.append_trace(wave = waves[i, :],
+                                     plaintext = text,
+                                     ciphertext = ciphertext,
+                                     key = key)
+
+                if capture_cfg.capture_mode == "aes_random":
+                    # Use ciphertext as next text, first text is the initial
+                    # one.
+                    text = ciphertext
+
+            # Compare received ciphertext with generated.
+            if capture_cfg.capture_mode == "aes_random":
+                compare_len = capture_cfg.output_len
+            else:
+                compare_len = 4
+            check_ciphertext(ot_aes, ciphertext, compare_len)
+
+            # Update the loop variable and the progress bar.
+            remaining_num_traces -= capture_cfg.num_segments
+            pbar.update(capture_cfg.num_segments)
+
+
+def print_plot(project: SCAProject, config: dict) -> None:
+    """ Print plot of traces.
+
+    Printing the plot helps to adjust the scope gain and check for clipping.
+
+    Args:
+        project: The project containing the traces.
+        config: The capture configuration.
+    """
+    if config["capture"]["show_plot"]:
+        plot.save_plot_to_file(project.get_waves(),
```

Comment:
are we loading all waves to RAM again here? :)

Reply:
Good point, thanks for the comment. I've changed the code - now only `num_traces `are loaded into memory.